addon/plugins/article-structure-plugin/commands/wrap-structure-content.ts

```diff
@@ -71,15 +69,120 @@ const wrapStructureContent = (
   };
 };
 
-function findInsertionContainer(selection: Selection, nodeType: NodeType) {
+function findInsertionContainer(
+  selection: Selection,
+  nodeType: NodeType,
+  schema: Schema,
+) {
   const { $from } = selection;
   for (let currentDepth = $from.depth; currentDepth >= 0; currentDepth--) {
     const currentAncestor = $from.node(currentDepth);
     const pos = currentDepth > 0 ? $from.before(currentDepth) : -1;
+
+    // Simple like for like replacement from 0th index
     if (
       currentAncestor.canReplaceWith(0, currentAncestor.childCount, nodeType)
     ) {
-      return { node: currentAncestor, depth: currentDepth, pos };
+      return {
+        node: currentAncestor.content,
+        from: pos + 1,
+        to: pos + currentAncestor.nodeSize - 1,
+      };
+    }
+
+    const currentAncestorParent = $from.node(currentDepth - 1);
+    const currentAncestorIndexInParent = $from.index(currentDepth - 1);
+
+    if (!currentAncestorParent) {
+      return null;
+    }
+
+    // Sometimes we might not be able to replace from 0th index, but we can try to go one
+    // level up and attempt to replace `currentAncestor` as a child of its parent
+    if (
+      currentAncestorParent.canReplaceWith(
+        currentAncestorIndexInParent,
+        currentAncestorIndexInParent + 1,
+        nodeType,
+      )
+    ) {
+      return {
+        node: currentAncestor,
+        from: pos,
+        to: pos + currentAncestor.nodeSize - 1,
+      };
+    }
+
+    // special case when we reach `doc` but cannot replace the content
+    // e.g. we reached doc, but there is "document_title article article", we might not be
+    // able to wrap the title around single `article` node, depending on the `content` definiton
+    // of the `doc`
+
+    /**
+     * Special case when we reach `doc` node, but cannot replace the content.
+     * e.g. we reached doc, and it contains "document_title article article",
+     * we might not be able to wrap single `article`, depending on the `content` definiton,
+     * of the `doc`. Therefore we need to find sibling nodes of same type and try
+     * to wrap them all together
+     */
+    if (currentAncestorParent.type.spec === schema.nodes.doc.spec) {
+      const doc = currentAncestorParent;
+
+      const start = pos <= 0 ? 0 : pos - 1;
+
+      const filter = ({ from }: { from: number }) => {
+        const node = doc.nodeAt(from);
+
+        return node?.type.name === currentAncestor.type.name;
+      };
+
+      const previousNodeIterator = findNodes({
```

Comment:
you mention you're looking for siblings, but findNodes loops through _every_ node
given you're looking for nodes with type of currentAncestor in currentAncestorParent, cant you just loop over currencAncestorParent.content?

Reply:
I don't think you need `.content`? the a `PNode` also supports `foreach` function which should do the exact same.

Reply:
true, even better, forgot that one existed :+1: